lib/level10/games/game.ex

```diff
@@ -5,46 +5,196 @@ defmodule Level10.Games.Game do
   game's state, and then stored on the server to be updated or to serve data
   down to clients.
   """
-
   alias Level10.Games.{Card, Player}
 
+  @type join_code :: <<_::32>>
   @type t :: %__MODULE__{
-          complete: boolean(),
-          current_round: integer(),
-          discard_pile: list(Card.t()),
-          draw_pile: list(Card.t()),
-          hands: map(),
-          players: map(),
-          scoring: map(),
-          table: map()
+          current_round: non_neg_integer() | :pending | :completed,
+          discard_pile: [Card.t()],
+          draw_pile: [Card.t()],
+          hands: %{optional(Player.id()) => [Card.t()]},
+          join_code: join_code(),
+          players: [Player.t()],
+          scoring: %{optional(Player.id()) => {non_neg_integer(), non_neg_integer()}},
+          table: %{optional(Player.id()) => keyword()}
         }
 
-  defstruct [
-    :complete,
-    :current_round,
-    :discard_pile,
-    :draw_pile,
-    :hands,
-    :players,
-    :scoring,
-    :table
-  ]
+  defstruct ~W[
+    current_round
+    discard_pile
+    draw_pile
+    hands
+    join_code
+    players
+    scoring
+    table
+  ]a
 
   @doc """
-  Takes a list of players and returns a new game struct with all values set to
-  defaults.
+  At the end of a round, the game struct should be passed into this function.
+  It will update player scoring and levels, check if the game has been
+  complete, and reset the state for the next round.
+
+  Make private later.
   """
-  @spec new(list(Player.t())) :: t()
-  def new(players) do
-    %__MODULE__{
-      complete: false,
-      current_round: 0,
+  @spec complete_round(t()) :: t()
+  def complete_round(game) do
+    game
+    |> update_scoring_and_levels()
+    |> check_complete()
+    |> clear_round()
+  end
+
+  @spec update_scoring_and_levels(t()) :: t()
+  defp update_scoring_and_levels(%{scoring: scoring, table: table, hands: hands} = game) do
+    scoring =
+      Map.new(scoring, fn {player, {level, score}} ->
```

Comment:
I like what you did here. I don't think I've ever used `Map.new/2` like this either so that's cool